metricbeat/module/windows/perfmon/pdh_windows_test.go

```diff
@@ -23,9 +23,86 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+var validQuery = `\Processor Information(_Total)\% Processor Time`
+
 // TestPdhErrno checks that PdhError provides the correct message for known
 // PDH errors and also falls back to Windows error messages for non-PDH errors.
 func TestPdhErrno_Error(t *testing.T) {
 	assert.Contains(t, PdhErrno(PDH_CSTATUS_BAD_COUNTERNAME).Error(), "Unable to parse the counter path.")
 	assert.Contains(t, PdhErrno(15).Error(), "The system cannot find the drive specified.")
 }
+
+// TestPdhOpenQueryInvalidQuery will check for file source and throw exception.
+func TestPdhOpenQueryInvalidQuery(t *testing.T) {
+	handle, err := PdhOpenQuery("invalid string", 0)
+	assert.EqualValues(t, handle, InvalidQueryHandle)
+	assert.EqualValues(t, err, PDH_FILE_NOT_FOUND)
+}
+
+// TestPdhAddCounterInvalidCounter checks for invalid query.
+func TestPdhAddCounterInvalidCounter(t *testing.T) {
+	handle, err := PdhAddCounter(InvalidQueryHandle, validQuery, 0)
+	assert.EqualValues(t, handle, InvalidCounterHandle)
+	assert.EqualValues(t, err, PDH_INVALID_ARGUMENT)
+}
+
+// TestPdhGetFormattedCounterValueInvalidCounter will test for invalid counters.
+func TestPdhGetFormattedCounterValueInvalidCounter(t *testing.T) {
+	counterType, counterValue, err := PdhGetFormattedCounterValue(InvalidCounterHandle, PdhFmtDouble)
+	assert.EqualValues(t, counterType, 0)
+	assert.EqualValues(t, counterValue, (*PdhCounterValue)(nil))
+	assert.EqualValues(t, err, PDH_INVALID_HANDLE)
+}
+
+// TestPdhExpandWildCardPathInvalidPath will test for invalid query path.
+func TestPdhExpandWildCardPathInvalidPath(t *testing.T) {
+	queryList, err := PdhExpandWildCardPath("sdfhsdhfd")
+	assert.Nil(t, queryList)
+	assert.EqualValues(t, err, PDH_INVALID_PATH)
+}
+
+// TestPdhCollectQueryDataInvalidQuery will check for invalid query.
+func TestPdhCollectQueryDataInvalidQuery(t *testing.T) {
+	err := PdhCollectQueryData(InvalidQueryHandle)
+	assert.EqualValues(t, err, PDH_INVALID_HANDLE)
+}
+
+// TestPdhCloseQueryInvalidQuery will check for invalid query.
+func TestPdhCloseQueryInvalidQuery(t *testing.T) {
+	err := PdhCloseQuery(InvalidQueryHandle)
+	assert.EqualValues(t, err, PDH_INVALID_HANDLE)
+}
+
+// TestPdhSuccessfulCounterRetrieval will execute the PDH  functions successfully.
+func TestPdhSuccessfulCounterRetrieval(t *testing.T) {
+	queryHandle, err := PdhOpenQuery("", 0)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer PdhCloseQuery(queryHandle)
+	queryList, err := PdhExpandWildCardPath(validQuery)
+	if err != nil {
+		t.Fatal(err)
+	}
+	queries := UTF16ToStringArray(queryList)
+	var counters []PdhCounterHandle
+	for _, query := range queries {
+		counterHandle, err := PdhAddCounter(queryHandle, query, 0)
+		if err != nil && err != PDH_NO_MORE_DATA {
+			t.Fatal(err)
+		}
+		counters = append(counters, counterHandle)
+	}
+	err = PdhCollectQueryData(queryHandle)
+	err = PdhCollectQueryData(queryHandle)
```

Comment:
Same here, is this needed twice here? Add a comment if it is.

Reply:
same as the above